app/controllers/user_sessions_controller.rb

```diff
@@ -117,7 +117,7 @@ def handle_social_login_flow(auth)
 
   def handle_site_login_flow
     username = params[:user_session][:username] if params[:user_session]
-    u = User.find_by(username: username)
+    u = User.find_by(username: username) || User.find_by(email: username)
```

Comment:
Kindly explain

Reply:
Oauth users are supposed be able to login with their email as username. So we need to check if the username they used to login is identical to an email in the db. That way, the code can check that `u` has a password. Otherwise, if a user logs in with their email,  `u` is nil and the code does not check if `u` has a password. If that happens, the correct error message is not displayed

Reply:
>  login with their email as username. 

Wrong. I have made a recursive hash for this. We don't use emails as usernames. Consider
bansal.sidharth@rediffmail.com
and bansal.sidharth@gmail.com
so I stored usernames bansal_sidharth and bansal_sidharth213 where 213 is some random hash created. Kindly check.

Reply:
No, what i mean is that oauth users can put their email in the username field when they login. For example, here:
https://github.com/publiclab/plots2/blob/d8c644a5304211ad461374d2bba4c90b25f86100/app/controllers/user_sessions_controller.rb#L131-L134

It checks if the username entered by the user is actually their email, and if this is the case and if their credentials are correct, they are logged in.

Reply:
Try to think regex of username and email are different. So, these queries are worthless. Does this makes sense?
cc\ @jywarren

Reply:
This isn't validation though. Currently, the login system allows oauth users to enter their email in the username field. It makes sense that if a user tries to log in with their email as username, they should get the same error messages as if they used their actual username.

Reply:
I am sorry, but I disagree here.
We can't treat emails as usernames in the db. 
> the login system allows oauth users to enter their email in the username field.

Oauth takes the email prefix as username. username has different regex. Username can't contain dots etc.  
Things are different.

Reply:
We don't treat the emails as usernames. We just check if the user entered their email instead of their username.

If this is not the expected behavior, other parts of the codebase that already work like this need to be changed

Reply:
> We can't treat emails as usernames in the db.

I think the difference is we're recognizing in the UI that people sometimes put their email instead of username, and the system adapts to accept that. It doesn't actually store the email as a username in the db, it just relabels the username input as an email during the auth process. Does that help clarify? 

It's a convenience but it does significantly reduce the incidence of people either forgetting their usernames, or just getting mixed up and using an email instead (which people do by mistake sometimes).

Reply:
>  It doesn't actually store the email as a username in the db

Ok. So, these changes 
> || User.find_by(email: username)

are not needed. @Uzay-G can you remove this change and see if test passes. I think test should pass.

Reply:
No it doesn't work if I do that. Instead the code continues and a `password invalid` error is displayed.

Reply:
I tried the test only passes with the changes

Reply:
Can you please explain what is happening at regex level?
If test is failing, we need to merge it. Hence, merging it.

Reply:
This is not going to break anything as changes are OR GATE. So, it is safe to merge. Hence merged it.

Reply:
This is not going to break anything as changes are OR GATE. So, it is safe to merge. Hence merged it.